usb/usb_host_hid/hid_host.c

```diff
@@ -0,0 +1,1474 @@
+            return true;
+        }
+    }
+    return false;
+}
+
+/**
+ * @brief HID Interface user callback function.
+ *
+ * @param[in] hid_iface   Pointer to an Interface structure
+ * @param[in] event_id    HID Interface event
+ */
+static inline void hid_host_user_interface_callback(hid_iface_t *hid_iface,
+        const hid_host_interface_event_t event)
+{
+    assert(hid_iface);
+
+    hid_host_dev_params_t *dev_params = &hid_iface->dev_params;
+
+    assert(dev_params);
+
+    if (hid_iface->user_cb) {
+        hid_iface->user_cb(hid_iface, event, hid_iface->user_cb_arg);
+    }
+}
+
+/**
+ * @brief HID Device user callback function.
+ *
+ * @param[in] event_id    HID Device event
+ * @param[in] dev_params  HID Device parameters
+ */
+static inline void hid_host_user_device_callback(hid_iface_t *hid_iface,
+        const hid_host_driver_event_t event)
+{
+    assert(hid_iface);
+
+    hid_host_dev_params_t *dev_params = &hid_iface->dev_params;
+
+    assert(dev_params);
+
+    if (s_hid_driver && s_hid_driver->user_cb) {
+        s_hid_driver->user_cb(hid_iface, event, s_hid_driver->user_arg);
+    }
+}
+
+/**
+ * @brief Add interface in a list
+ *
+ * @param[in] hid_device    HID device handle
+ * @param[in] iface_desc  Pointer to an Interface descriptor
+ * @param[in] hid_desc    Pointer to an HID device descriptor
+ * @param[in] ep_desc     Pointer to an EP descriptor
+ * @return esp_err_t
+ */
+static esp_err_t hid_host_add_interface(hid_device_t *hid_device,
+                                        const usb_intf_desc_t *iface_desc,
+                                        const hid_descriptor_t *hid_desc,
+                                        const usb_ep_desc_t *ep_in_desc)
+{
+    hid_iface_t *hid_iface = calloc(1, sizeof(hid_iface_t));
+
+    HID_RETURN_ON_FALSE(hid_iface,
+                        ESP_ERR_NO_MEM,
+                        "Unable to allocate memory");
+
+    HID_ENTER_CRITICAL();
+    hid_iface->parent = hid_device;
+    hid_iface->state = HID_INTERFACE_STATE_NOT_INITIALIZED;
+    hid_iface->dev_params.addr = hid_device->dev_addr;
+
+    if (iface_desc) {
+        hid_iface->dev_params.iface_num = iface_desc->bInterfaceNumber;
+        hid_iface->dev_params.sub_class = iface_desc->bInterfaceSubClass;
+        hid_iface->dev_params.proto = iface_desc->bInterfaceProtocol;
+    }
+
+    if (hid_desc) {
+        hid_iface->country_code = hid_desc->bCountryCode;
+        hid_iface->report_desc_size = hid_desc->wReportDescriptorLength;
+    }
+
+    // EP IN && INT Type
+    if (ep_in_desc) {
+        if ( (ep_in_desc->bEndpointAddress & USB_B_ENDPOINT_ADDRESS_EP_DIR_MASK) &&
+                (ep_in_desc->bmAttributes & USB_B_ENDPOINT_ADDRESS_EP_NUM_MASK) ) {
+            hid_iface->ep_in = ep_in_desc->bEndpointAddress;
+            hid_iface->ep_in_mps = USB_EP_DESC_GET_MPS(ep_in_desc);
+        } else {
+            ESP_LOGE(TAG, "HID device EP IN %#X configuration error",
```

Comment:
We're in a critical section here, so use `ESP_EARLY_LOGx` here instead. `ESP_LOGx` is a blocking API, so we'll usually get an assert you attempt to block in a critical section.

Reply:
Exactly! 
Thanks.